xcube/util/timeindex.py

```diff
@@ -0,0 +1,141 @@
+    it with a modified value for labels['time'].
+
+    If there is no 'time' key in the labels dictionary or if there is no
+    'time' dimension in the var array, the original labels are returned.
+
+    If there is a 'time' key in the labels dictionary and a 'time' dimension
+    in the var array, they are checked for compatibility. If they are
+    compatible, the original labels are returned. If not, an altered labels
+    dictionary is returned, in which the time key has been modified to be
+    compatible with the type of the 'time' dimension in the var array.
+
+    See the documentation for `ensure_time_index_compatible` for details on
+    the compatibility check.
+    """
+
+    if 'time' in labels and 'time' in var.dims:
+        return dict(labels,
+                    time=ensure_time_index_compatible(var, labels['time']))
+    else:
+        return labels
+
+
+def ensure_time_index_compatible(var: Union[xr.DataArray, xr.Dataset],
+                                 timeval: Any) -> Any:
+    """Ensure that timeval is a valid time indexer for var
+
+    It is expected that the value of timeval will be a valid timestamp (i.e.
+    a valid input to pandas.Timestamp.__init__), or a slice in which the
+    start and stop fields are valid timestamps. For a slice, the start and
+    stop fields are processed separately, and their modified values (if
+    required) are returned as the start and stop fields of a new slice. The
+    step field is included unchanged in the new slice.
+
+    The compatibility check consists of checking the timezone-awareness
+    of the variable and its indexer. If the variable is timezone-aware and
+    the indexer is timezone-naive, or vice versa, a new indexer is returned
+    whose timezone-awareness corresponds to that of the variable. Otherwise
+    the original indexer is returned.
+    """
+    
+    if isinstance(timeval, slice):
+        # process start and stop separately, and pass step through unchanged
+        return slice(
+            _ensure_timestamp_compatible(var, timeval.start),
+            _ensure_timestamp_compatible(var, timeval.stop),
+            timeval.step)
+    else:
+        return _ensure_timestamp_compatible(var, timeval)
+
+
+def _ensure_timestamp_compatible(var: xr.DataArray, timeval: Any):
+    if timeval is None:
+        return None
+
+    if hasattr(timeval, "tzinfo"):
+        timestamp = timeval
+        timeval_tzinfo = timeval.tzinfo
+    else:
+        try:
+            timestamp = pd.Timestamp(timeval)
+            timeval_tzinfo = timestamp.tzinfo
+        except (TypeError, ValueError):
+            logger.warning("Can't determine indexer timezone, leaving it "
+                           "unmodified.")
+            return timeval
+
+    if _has_datetime64_time(var):
+        # pandas treats all datetime64 arrays as timezone-naive
+        array_timezone = None
+    elif hasattr(var.time.values[0], 'tzinfo'):
+        array_timezone = var.time.values[0].tzinfo
```

Comment:
```suggestion
        array_timezone = var.time[0].values.tzinfo
```

Reply:
Rather surprisingly, this breaks the test, because indexing time like that causes xarray to automatically convert the value to a `datetime64` with a timezone (!). But I found a workaround, and will implement it here and in the line above:

``` python
var.time[0:1].values[0].tzinfo
```

Looks a little odd, but slicing the time rather than indexing it preserves the initial datatype, and avoids pulling the whole time array into memory.